YatimaStdLib/Nat.lean

```diff
@@ -103,6 +103,63 @@ theorem div2_lt (h : n ≠ 0) : n / 2 < n := by
     exact @div2_lt (n + 2) (by simp_arith)
     simp_arith
 
+/--
+Evaluates `b^e mod m`
+-/
+def powMod (m b e : Nat) : Nat :=
+  let rec go (b e r : Nat) : Nat :=
+    if h : e = 0 then r
+    else
+      let e' := e / 2
+      have : e' < e := 
+      by exact Nat.div2_lt h
+      if e % 2 = 0
+      then go ((b*b) % m) e' r              -- TODO : Use Montgomery multiplication here to avoid
+      else go ((b*b) % m) e' ((r*b) % m)    --        calculating `mod` at every step
+  go b e 1
+
+/-- A legendre symbol denotes the value of `a^((p - 1)/2) mod p`
+-/
+def legendre (a : Nat) (p : Nat) : Nat :=  -- TODO : Use a pre-calculated `(p - 1) / 2 ` AddChain
+  powMod p a ((p - 1) / 2)                 --        and AddChain `fastExp` here
+
+/--
+The Tonelli-Shanks algorithm solves the equation having the form
+`x^2 = n mod p`, whenever it exists
+-/
```

Comment:
It's good to link the original Python implementation here

Reply:
Done